src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -694,13 +689,44 @@ impl RustcTargetData {
             target_config.insert(ct, host_config.clone());
         }
 
-        Ok(RustcTargetData {
+        let mut res = RustcTargetData {
             rustc,
+            config,
+            requested_kinds: requested_kinds.into(),
             target_config,
             target_info,
             host_config,
             host_info,
-        })
+        };
+
+        // Get all kinds we currently know about. Further kinds will
+        // be added via the `add_kind` function below.
+        let all_kinds = requested_kinds
+            .iter()
+            .copied()
+            .chain(ws.members().flat_map(|p| {
```

Comment:
My comments later about this needing to traverse the whole dependency graph I don't think are quite right until we have artifact dependencies. For now I think it should be safe to only traverse the workspace since that's the only location of the root units which is where targets can come from. Can you add a comment to this effect here though?